### PR TITLE
UMAP and UMAP Plot Tests Refactoring

### PR DESCRIPTION
**Pull request by leriomaggio**
## Overview

This PR brings a complete **refactoring** of the UMAP test suites for easier testability and code maintainability.

As a minor change, `pandas` has been also included in the dependencies required to work with `umap.plot`.

## Details

This PR brings a complete **refactoring** of the UMAP `tests` packages to ease testing of different UMAP functionalities, to allow for (partial) tests execution, and for easier integration and maintainability. 

The test suite has been re-organised in multiple testing modules,
including test cases related to a specific
set of UMAP features under test.
(All tests where previously **all** included in the `test_umap.py` module.

**`pytest`** is the reference backend for testing environment and execution,
also integrating with pre-existent nose-based tests.
(_More on the resulting advantages, later on_)

Whenever needed, the code of each test case has been refactored in 
order to combine shared (and repeated)
testing operations in  _utility_ testing functions, defined on top of each test module.
This has the advantage of centralising the testing code (single point of failure), 
and to avoid code clones (repetitions), resulting In shorter testing code 
easier to maintain.

All data dependency has been implemented as **test fixtures** (`pytest.fixture`) 
taking over shared global variables.
All the fixtures that are needed by test cases in multiple modules have been 
centralised in the pytest (default) `conftest.py` module.

This data injection mechanism based on **fixtures** allows
the execution of each test modules in isolation, as well
as within the whole test suite.
`pytest` will inject the dependencies automatically.

The list of Test modules (as a result of the refactoring) is:

- `test_plot`: basic tests for umap.plot
- `test_umap_df_validation_params`: Tests on parameters validation for DataFrameUMAP
- `test_umap_metrics`: Tests for UMAP metrics - spatial, binary, and sparse
- `test_umap_nn`: Tests for NearestNeighbours
- `test_umap_on_iris`: Tests for UMAP on Iris Dataset
- `test_umap_ops`: Tests for general UMAP ops (e.g. clusterability, transform stability)
- `test_umap_repeated_data`: UMAP tests on repeated data (sparse|dense; spatial|binary)
- `test_umap_trustworthiness`: Tests on UMAP Trustworthiness
- `test_umap_validation_params`: Tests for fit parameters validation

### Notes:

For tests in the old `test_umap.py`: NO new test has been **added** (yet? 🤞 on the PR 😄 ) - only re-organised in multiple modules, removing code clones (duplications) in testing code, and setting up `fixtures` - so removing globe variables.

As for `test_plot`, a single **new** test has been added to check that an exception is raised in case the environment is **not** correctly setup (using `unittest.mock`).

### CODE REVIEW

**Comment by leriomaggio** [comment]
First of all, apologies for the _many_ commit after the initial PR: It took a little while for me to figure out who to **actually** change `pytest` as the actual testing backend. 

To do this, I also changed then Travis, Azure Pipeline, and AppVeyor pipelines configuration files. 
Those now are also part of the PR.

In addition (see this [commit](https://github.com/lmcinnes/umap/pull/394/commits/91adcb53815afe0cec569d69700a64b517a4e43f) I also updated the integration with the `trustworthiness` function import, making it compliant to current (and future)  `sklearn.manifold` API.

**Comment by lmcinnes** [comment]
Thanks for al the work you have obviously put it. It is going to take me a while to go through and properly digest this, but it looks comprehensive enough that I want to manage to wrap my head around everything  this is doing properly before I merge it. That means it is going to take me a little time, so please bear with me if I am a little slow -- it isn't for lack of appreciation or enthusiasm for your work; simply my time availability and managing to get through it all.

**Comment by leriomaggio** [comment]
Hi Leland @lmcinnes, 
    thank you very much for your reply, and for being willing to review the PR.

No worries at all for the slow replies. It makes total sense. 
Besides, please feel free to comment here for anything that is not clear enough. 

The only bit I would like to remark though is: **no new test** has been added (a part for a couple ones in `test_plot`, but they're just to check the environment - don't really count). 
So tests have just been re-organised - and fixtures have been used to avoid (shared) globals.
So, I hope this will ease the review process.

Truth to be told, the whole story behind this PR si that it actually came as a "consequence" of another (couple) of PR I am working on. 

I started thinking of generating some tests to check that my intuition was right.. and here I am 😄 

Looking forward to your comments, and thanks again for your work on this.

**Comment by lmcinnes** [comment]
Thanks -- this looks really good and will go a long way to making it easier to build out testing from here. I agree that the plotting testing is ... not great, but I am unsure of best practices for testing plotting functions (perhaps I should look at how Yellowbrick does things).
